File: src/main/java/org/mitallast/queue/rest/netty/HttpServerInitializer.java
package org.mitallast.queue.rest.netty;

import io.netty.channel.Channel;
import io.netty.channel.ChannelInitializer;
import io.netty.channel.ChannelPipeline;
import io.netty.handler.codec.http.HttpObjectAggregator;
import io.netty.handler.codec.http.HttpServerCodec;
import io.netty.handler.stream.ChunkedWriteHandler;

public class HttpServerInitializer extends ChannelInitializer {

    private final HttpServerHandler httpHandler;
    private final WebSocketFrameHandler webSocketFrameHandler;

    public HttpServerInitializer(HttpServerHandler httpHandler, WebSocketFrameHandler webSocketFrameHandler) {
        super();
        this.httpHandler = httpHandler;
        this.webSocketFrameHandler = webSocketFrameHandler;
    }

    @Override
    protected void initChannel(Channel ch) throws Exception {
        ChannelPipeline pipeline = ch.pipeline();
        pipeline.addLast(new HttpServerCodec(4096, 8192, 8192, false));
        pipeline.addLast(new HttpObjectAggregator(65536));
        pipeline.addLast(new ChunkedWriteHandler());
//        pipeline.addLast(new WebSocketServerCompressionHandler());
//        pipeline.addLast(new WebSocketServerProtocolHandler("/ws/", null, true));
//        pipeline.addLast(webSocketFrameHandler);
        pipeline.addLast(httpHandler);
    }
}

File: src/main/java/org/mitallast/queue/rest/netty/HttpServerInitializer.kt
package org.mitallast.queue.rest.netty

import io.netty.channel.ChannelInitializer
import io.netty.channel.socket.SocketChannel
import io.netty.handler.codec.http.HttpObjectAggregator
import io.netty.handler.codec.http.HttpServerCodec
import io.netty.handler.codec.http.websocketx.WebSocketServerProtocolHandler
import io.netty.handler.codec.http.websocketx.extensions.compression.WebSocketServerCompressionHandler
import io.netty.handler.stream.ChunkedWriteHandler
import io.netty.util.concurrent.DefaultEventExecutorGroup
import io.netty.util.concurrent.DefaultThreadFactory

class HttpServerInitializer(
    private val httpHandler: HttpServerHandler,
    private val webSocketFrameHandler: WebSocketFrameHandler
) : ChannelInitializer<SocketChannel>() {

    private val group = DefaultEventExecutorGroup(8, DefaultThreadFactory("http handler"));

    override fun initChannel(ch: SocketChannel) {
        val pipeline = ch.pipeline()
        pipeline.addLast(HttpServerCodec(4096, 8192, 8192, false))
        pipeline.addLast(HttpObjectAggregator(65536))
        pipeline.addLast(ChunkedWriteHandler())
        pipeline.addLast(WebSocketServerCompressionHandler())
        pipeline.addLast(WebSocketServerProtocolHandler("/ws/", null, true))
        pipeline.addLast(webSocketFrameHandler)
        pipeline.addLast(group, httpHandler)
    }
}